design.md

```diff
@@ -0,0 +1,306 @@
+# Design
+
+This document outlines the design of the systems within this repo.
+
+## Design Approach
+
+There's two main aspects to my approach I tried to maintain within each package.
+1. Have critical functionality hang off a `Service` type. This allows me to minimize global state, create a recognizable pattern for readers, and utilize interfaces if needed.
+2. Always take the simplest approach. It was made clear in the challenge description a high performing nor a scalable system was being asked for. Therefore, I designed components in a manner I know will not scale in order to keep things simple. More on this in [Performance Concerns](#performance-concerns).
+
+
+## API
+
+Click [here](https://github.com/tjper/teleport/blob/design/proto/jobworker/v1/service_api.proto) for API definition.
+
+## Authentication
+
+Client and server will utilize mTLS.
+
+TLS 1.3 will be used to ensure a secure cipher suite is utilized. As of Go 1.17, the cipher suites are ordered by `crypto/tls` based on range of factors. https://go.dev/blog/tls-cipher-suites
+
+A set of certificates and secrets will exist in the `certs/` directory. All certificates will be signed by the included CA.
```

Comment:
Please mention the certificate parameters too.

Reply:
I've updated this. Would you also like to see parameters like ...
- organization
- country
- locality
- organizational-unit
- province
- ip
- uri

Reply:
That's OK, we don't need all of that in design. I was mostly looking for algorithm/key length.